api_app/analyzers_manager/observable_analyzers/greynoiseintel.py

```diff
@@ -0,0 +1,58 @@
+# This file is a part of IntelOwl https://github.com/intelowlproject/IntelOwl
+# See the file 'LICENSE' for copying permission.
+
+from greynoise import GreyNoise
+
+from api_app.analyzers_manager import classes
+from api_app.exceptions import AnalyzerRunException
+from tests.mock_utils import MockResponse, if_mock_connections, patch
+
+
+class GreyNoiseAnalyzer(classes.ObservableAnalyzer):
+    def set_params(self, params):
+        self.api_version = params.get("greynoise_api_version", "v3")
+        self.max_records_to_retrieve = int(params.get("max_records_to_retrieve", 500))
+
+    def run(self):
+
+        if self.api_version == "v2":
+            api_key = self._secrets["api_key_name"]
+            session = GreyNoise(
+                api_key=api_key, integration_name="greynoise-intelowl-v1.0"
+            )
+            context_response = session.ip(self.observable_name)
+            riot_response = session.riot(self.observable_name)
+            response = context_response.copy()
+            response.update(riot_response)
+
+        elif self.api_version == "v3":
+            api_key = self._secrets["api_key_name"]
+            session = GreyNoise(
+                api_key=api_key,
+                integration_name="greynoise-community-intelowl-v1.0",
+                offering="Community",
+            )
+            response = session.ip(self.observable_name)
+
+        else:
+            raise AnalyzerRunException(
+                "Invalid API Version. " "Supported are: v2 (paid), v3 (community)"
+            )
+
+        return response
+
+    @classmethod
+    def _monkeypatch(cls):
+        patches = [
+            if_mock_connections(
+                patch(
+                    "requests.get",
+                    return_value=MockResponse({}, 200),
+                ),
+                patch(
+                    "requests.post",
+                    return_value=MockResponse({}, 200),
+                ),
+            )
+        ]
+        return super()._monkeypatch(patches=patches)
```

Comment:
for the tests you should just mock the Classes/functions that you call in your code here with the results that you want.

This is an example: https://github.com/intelowlproject/IntelOwl/blob/master/api_app/analyzers_manager/file_analyzers/cape_sandbox.py#L209

Reply:
so you should mock `greynoise.Greynoise.<methods>` and add some possible output that you would get in real life

Reply:
@mlodic : I'm having an issue trying to make this work, I think I just don't have enough familiarity with this functionality.  This is what I've been trying, but I can't get it to work.  Any guidance you can provide?

```
    @classmethod
    def _monkeypatch(cls):
        patches = [
            if_mock_connections(
                patch(
                    "greynoise.GreyNoise.ip",
                    return_value=MockResponse(
                        {"noise": True}, 200
                    ),
                ),
                patch(
                    "greynoise.GreyNoise.riot",
                    return_value=MockResponse(
                        {"riot": True}, 200
                    ),
                ),
            )
        ]
        return super()._monkeypatch(patches=patches)
```

I keep facing this and I'm not sure how to work around it: `[('GreyNoise', ["GreyNoise Exception: unsupported operand type(s) for |=: 'MockResponse' and 'MockResponse'"])]`

Reply:
I wanted to fix this for you but I cannot push the changes in this PR, I don't know why exactly, maybe some permissions on your fork side...
try this please:


```
   @classmethod
    def _monkeypatch(cls):
        patches = [
            if_mock_connections(
                patch(
                    "greynoise.Greynoise",
                    return_value=MockGreynoise(),
                ),
            )
        ]
        return super()._monkeypatch(patches=patches)```

and add this class on top of the module:

`class MockGreynoise:
    def __init__(self):
        pass

    def ip(self, *args, **kwargs):
        return {"noise": True}

    def riot(self, *args, **kwargs):
        return {"riot": True}`

Reply:
@mlodic : tried this, but now I'm back to getting a 401.  Would you like me to push the change up to the branch so you can see if I did anything wrong?

Reply:
@mlodic : alright, latest commit is pushed so you can see where I'm at.  Appreciate the help.